code/graph_algorithms/src/centroid_decomposition/centroid_decompostition.cpp

```diff
@@ -0,0 +1,117 @@
+#include <iostream>
+#include <vector>
+
+std::vector<std::vector<int>> Level;
+
+class Graph 
+{
+public:
+    int ver_;
+    std::vector<std::vector<int>> adj_;
+    std::vector<std::vector<int>> centroidTree_;
+    std::vector<int> sizes_;
+    std::vector<bool> marked_;
+    Graph(int num) :
+        ver_(num),
+        adj_(num + 1),
+        centroidTree_(num + 1),
+        sizes_(num + 1),
+        marked_(num + 1)
+        {}
+
+    void addEdge(int x, int y, std::vector<std::vector<int>> &graph);
+    void calcSizes(int cur, int par, std::vector<std::vector<int>> graph);
+    int findCentroid(int cur, int par, int vertices, std::vector<std::vector<int>> graph);
+    void decomposeTree(int cur, int par, int total, int level, std::vector<std::vector<int>> graph);
+};
+
+void Graph::addEdge(int x, int y, std::vector<std::vector<int>> &graph)
+{
+    if (x < 0 || x > ver_)
+    {
+        throw std::out_of_range{"x is out of boundaries"};
+    }
+    if (y < 0 || y > ver_)
+    {
+        throw std::out_of_range{"y is out of boundaries"};
+    }
+    graph[x].push_back(y);
+    graph[y].push_back(x);
+}
+
+void Graph::calcSizes(int cur, int par, std::vector<std::vector<int>> graph)
+{
+    sizes_[cur] = 1;
+    for (const auto& to : graph[cur])
+    {
+        if (!(to == par || marked_[to] == true))
+        {
+            calcSizes (to, cur, graph);
+            sizes_[cur] += sizes_[to];
+        }
+    }
+}
+
+int Graph::findCentroid(int cur, int par, int vertices, std::vector<std::vector<int>> graph){
+    for (const auto& to : graph[cur])
+    {
+        if (!(to == par || marked_[to] == true))
+        {
+            if (sizes_[to] > vertices / 2)
+            {
+                return findCentroid(to, cur, vertices, graph);
+            }
+        }
+        return cur;
+    }
+}
+
+void Graph::decomposeTree(int cur, int par, int total, int level, std::vector<std::vector<int>> graph)
+{
+    calcSizes(cur, -1, graph);
+    int centroid = findCentroid(cur, -1, sizes_[cur], graph);
+    Level[level].push_back(centroid);
+    calcSizes(centroid, -1, graph);
+    marked_[centroid] = true;
+    for (const auto& to : adj_[centroid])
+    {
+        if(!(to == par || marked_[to] == true))
+        {
+            decomposeTree(to, cur, sizes_[to], level + 1, graph);
+            addEdge(cur, to, centroidTree_);
+        }
+    }
+}
+
+void showOutput()
+{
+    for (int i = 1; Level[i].empty() == false; ++i)
+    {
+        std::cout << "Centroids of level " << i << ": ";
+        bool more_than_1 = false;
+        for(const auto& j : Level[i])
+        {
+            if (more_than_1 == true)
+            {
+                std::cout << ", ";
+            }
+            std::cout << j;
+            more_than_1 = true;
+        }
+        std::cout << '\n';
+    }
+}
+
+int main(){
+    Graph tree(7);
+    Level.resize(8);
+    tree.addEdge(1, 2, tree.adj_);
```

Comment:
Since you are trying to pass member variables to the tree, I would just remove the parameter and directly modify these variables inside addEdge. Then you can make the variables private again.

Reply:
@arnavb But there the moment where I add edges in another tree (Centroid Tree) using addEdge. If I modify it in the function, then how am I going to add edges to Centroid Tree?